### DISCUSSION TITLE
Forward Hessian issue

### DISCUSSION THREAD

**Discussion by Manushanker** [autodiff/autodiff#299] [General]
Hi Allan
I am using the forward/dual hessian function to obtain the Jacobian and Hessian matrix of a user defined function. Found that the hessian function calls the specified function for computing first and second order derivatives for every variable. May I know if it is possible to compute the Jacobian and Hessian by a single function call?

Thanks a lot for your time and help.

**Comment by allanleal** [comment]
Hi @Manushanker, this is not possible with `dual` and `real` which track functional dependencies against a single variable each time. What you can do is calculate the Jacobian and/or Hessian with respect to some selected variables; those who contribute most significantly to the matrix. You would then need to use the `seed` method before the function evaluation and then extract the derivatives with respect to the seeded variable. Unseed the current seeded variable and repeat the process for another variable of interest. Check the implementation of the `jacobian` method and you will see how this is done.

**Reply by Manushanker** [reply]
Hi Allan,

Thanks for your reply! The process you explained seems very complicated. If possible, can you please implement a new Jacobian/Hessian function based on the seed method? I am using your library to generate results for a journal paper. The computation time required for obtaining the first order gradients is far superior to other libraries such as CoDIPack, CPPAD. It would be excellent if you could implement a new Hessian function based on the seed method.

Thanks for your help.

**Comment by allanleal** [comment]
Are you using type var? This type has not been optimized. Due to other
projects, I have mainly invested time improving types real and dual. Type
real is what I use mostly in my research projects.

**Reply by Manushanker** [reply]
I am not using type var to compute the gradients as it is slower than real or dual. I used the 'derivativesx(u, wrt(x, y, z));' described in the example 'examples/reverse/example-reverse-higher-order-derivatives-multi-variable-function.cpp' to determine the elements in the Hessian matrix. This method is very very slow and as you said it is not optimized as real type. Is there a implementation of derivativesx function for real type?

**Reply by Manushanker** [reply]
HI Allan,

Can you please let me know if you are planning to implement Hessian function with a single functional call? 

Thanks.

Manushanker

**Reply by allanleal** [reply]
Hello Manushanker, I wish I could implement what you want, but unfortunately this is not feasible for me at the moment if there is no funding for it. The developments that go into autodiff are mainly based on what I need for my main research project, https://reaktoro.org, on simulating chemical reaction problems. For that, I mainly use `real` (first order derivatives). It's the type I've found that works best for my needs after developing the other types and trying them out.

**Reply by Manushanker** [reply]
Thanks for your reply!
